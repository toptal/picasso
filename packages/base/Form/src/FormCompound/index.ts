import { Form } from '../Form'
import { FormError } from '../FormError'
import { FormField } from '../FormField'
import { FormHint } from '../FormHint'
import { FormLabel } from '../FormLabel'

type FormCompoundType = typeof Form & {
  Field: typeof FormField
  Hint: typeof FormHint
  Error: typeof FormError
  Label: typeof FormLabel
}

export const FormCompound: FormCompoundType = Object.assign(Form, {
  Field: FormField,
  Hint: FormHint,
  Error: FormError,
  Label: FormLabel,
})
