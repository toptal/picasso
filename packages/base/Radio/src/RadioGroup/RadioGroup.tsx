/* eslint-disable react/no-array-index-key */
import React from 'react'
import type { RadioGroupProps } from '@material-ui/core'
import { RadioGroup as MUIRadioGroup } from '@material-ui/core'
import { useTheme } from '@material-ui/core/styles'
import type { GridSizeProps, GridProps } from '@toptal/picasso-grid'
import { GridCompound as Grid } from '@toptal/picasso-grid'

type GridSpacing = GridProps['spacing']

export interface Props extends RadioGroupProps, GridSizeProps {
  /** Align radios horizontally  */
  horizontal?: boolean
  /** Defines amount of space between radio buttons components (in px) */
  spacing?: GridSpacing
}

const RadioGroup = (props: Props) => {
  const { horizontal, spacing, xs, sm, md, lg, xl, ...rest } = props
  const { spacing: themeSpacing } = useTheme()

  const direction = horizontal ? 'row' : 'column'
  const gridSpacing = spacing ?? horizontal ? themeSpacing(2) : 0

  const children = React.Children.toArray(rest.children)

  return (
    <MUIRadioGroup {...rest}>
      <Grid
        direction={direction}
        spacing={gridSpacing as GridSpacing}
        className='mt-0 mb-0'
      >
        {children.map((child, index) => (
          <Grid.Item
            key={index}
            className='leading-none [&&]:pt-0 [&&]:pb-0'
            xs={xs}
            sm={sm}
            md={md}
            lg={lg}
            xl={xl}
          >
            {child}
          </Grid.Item>
        ))}
      </Grid>
    </MUIRadioGroup>
  )
}

RadioGroup.defaultProps = {
  horizontal: false,
}

export default RadioGroup
