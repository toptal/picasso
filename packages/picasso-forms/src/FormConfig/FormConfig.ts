import { createContext, useContext } from 'react'

export type RequiredVariant = 'default' | 'asterisk'

type ValidationOnSubmitConfig = {
  showValidState?: false
  validateOnSubmit?: true
}

type ShowValidStateConfig = {
  showValidState?: true
  validateOnSubmit?: false
}

type NoValidationConfig = {
  showValidState?: false
  validateOnSubmit?: false
}

// reason for different types is that we want to deny the user to set both showValidState and validateOnSubmit to true
export type FormConfigProps = (
  | ValidationOnSubmitConfig
  | ShowValidStateConfig
  | NoValidationConfig
) & {
  requiredVariant?: RequiredVariant
  highlightAutofill?: boolean
}

export const FormConfigContext = createContext<FormConfigProps>({
  highlightAutofill: false,
})
export const FormConfigProvider = FormConfigContext.Provider

export const useFormConfig = () => useContext(FormConfigContext)
