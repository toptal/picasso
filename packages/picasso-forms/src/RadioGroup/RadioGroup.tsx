import React from 'react'
import {
  Radio as PicassoRadio,
  RadioProps,
  RadioGroupProps,
} from '@toptal/picasso'

import PicassoField, { FieldProps } from '../Field'
import FieldLabel from '../FieldLabel'
import RadioGroupContext from './RadioGroupContext'

export type Props = RadioGroupProps & FieldProps<RadioProps['value']>

export const RadioGroup = (props: Props) => {
  const { children, label, titleCase, ...rest } = props

  return (
    <RadioGroupContext.Provider value={props.name}>
      <PicassoField
        {...rest}
        type='radio'
        label={
          label ? (
            <FieldLabel
              name={props.name}
              required={props.required}
              label={label}
              titleCase={titleCase}
            />
          ) : null
        }
      >
        {radioGroupProps => (
          <PicassoRadio.Group {...radioGroupProps}>
            {children}
          </PicassoRadio.Group>
        )}
      </PicassoField>
    </RadioGroupContext.Provider>
  )
}

RadioGroup.defaultProps = {}

RadioGroup.displayName = 'RadioGroup'

export default RadioGroup
