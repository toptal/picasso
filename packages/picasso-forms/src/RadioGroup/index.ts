export { default } from './RadioGroup'
export { default as RadioGroupContext } from './RadioGroupContext'
export type { Props } from './RadioGroup'
