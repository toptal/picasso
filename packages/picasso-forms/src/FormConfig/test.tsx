import React from 'react'
import { screen, render, fireEvent } from '@toptal/picasso/test-utils'

import { FormCompound as Form } from '../FormCompound'
import { FormConfigProps } from './FormConfig'

const TEXT_INPUT_LABEL = 'Test text field'

const renderForm = (configProps: FormConfigProps) => {
  return render(
    <Form.ConfigProvider value={configProps}>
      <Form onSubmit={() => {}}>
        <Form.Input label={TEXT_INPUT_LABEL} required name='test' />
        <Form.SubmitButton>Submit</Form.SubmitButton>
      </Form>
    </Form.ConfigProvider>
  )
}

describe('Form.ConfigProvider', () => {
  it('validate only on submit', async () => {
    renderForm({ validateOnSubmit: true })

    fireEvent.blur(screen.getByLabelText(TEXT_INPUT_LABEL))

    expect(
      screen.queryByText('Please complete this field.')
    ).not.toBeInTheDocument()

    fireEvent.click(screen.getByRole('button', { name: 'Submit' }))

    expect(
      await screen.findByText('Please complete this field.')
    ).toBeInTheDocument()
  })

  it('validate normally on blur / change', async () => {
    renderForm({ validateOnSubmit: false })

    fireEvent.blur(screen.getByLabelText(TEXT_INPUT_LABEL))

    expect(screen.getByText('Please complete this field.')).toBeInTheDocument()
  })
})
