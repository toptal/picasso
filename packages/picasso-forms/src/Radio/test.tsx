import React from 'react'
import { render } from '@toptal/picasso/test-utils'

import { FormConfigProps } from '../FormConfig'
import { FormCompound as Form } from '../FormCompound'
import { Props as RadioGroupProps } from '../RadioGroup'

const renderFormRadio = (
  { required, name }: RadioGroupProps,
  formConfig: FormConfigProps = {}
) =>
  render(
    <Form.ConfigProvider value={formConfig}>
      <Form onSubmit={() => {}} initialValues={{ color: '#ffe4b5' }}>
        <Form.RadioGroup
          name={name}
          label="What's your favorite color?"
          required={required}
        >
          <Form.Radio label='Crimson' value='#ed143d' />
          <Form.Radio label='Moccasin' value='#ffe4b5' />
        </Form.RadioGroup>
      </Form>
    </Form.ConfigProvider>
  )

describe('FormRadio', () => {
  it('renders', () => {
    const { container } = renderFormRadio({
      name: 'color',
    })

    expect(container).toMatchSnapshot()
  })

  it('required with asterisk', () => {
    const { container } = renderFormRadio(
      {
        name: 'color',
        required: true,
      },
      {
        requiredVariant: 'asterisk',
      }
    )

    expect(container).toMatchSnapshot()
  })
})
