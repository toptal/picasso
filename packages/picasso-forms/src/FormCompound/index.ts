import { FieldRequirements } from '@toptal/picasso'

import Form from '../Form'
import Autocomplete from '../Autocomplete'
import Input from '../Input'
import Select from '../Select'
import Radio from '../Radio'
import ButtonRadio from '../ButtonRadio'
import RadioGroup from '../RadioGroup'
import Checkbox from '../Checkbox'
import ButtonCheckbox from '../ButtonCheckbox'
import CheckboxGroup from '../CheckboxGroup'
import NumberInput from '../NumberInput'
import FileInput from '../FileInput'
import DatePicker from '../DatePicker'
import TimePicker from '../TimePicker'
import TagSelector from '../TagSelector'
import SubmitButton from '../SubmitButton'
import Switch from '../Switch'
import Rating from '../Rating'
import Dropzone from '../Dropzone'
import RichTextEditor from '../RichTextEditor'
import PasswordInput from '../PasswordInput'
import { FormConfigContext } from '../FormConfig'
import AvatarUpload from '../AvatarUpload'

type FormCompoundType = typeof Form & {
  Autocomplete: typeof Autocomplete
  Input: typeof Input
  Select: typeof Select
  Radio: typeof Radio
  ButtonRadio: typeof ButtonRadio
  RadioGroup: typeof RadioGroup
  Checkbox: typeof Checkbox
  ButtonCheckbox: typeof ButtonCheckbox
  CheckboxGroup: typeof CheckboxGroup
  NumberInput: typeof NumberInput
  FileInput: typeof FileInput
  DatePicker: typeof DatePicker
  TimePicker: typeof TimePicker
  TagSelector: typeof TagSelector
  SubmitButton: typeof SubmitButton
  ConfigProvider: typeof FormConfigContext.Provider
  Switch: typeof Switch
  Rating: typeof Rating
  Dropzone: typeof Dropzone
  PasswordInput: typeof PasswordInput
  FieldRequirements: typeof FieldRequirements
  RichTextEditor: typeof RichTextEditor
  AvatarUpload: typeof AvatarUpload
}

export const FormCompound: FormCompoundType = Object.assign(Form, {
  Autocomplete: Autocomplete,
  Input: Input,
  Select: Select,
  Radio: Radio,
  ButtonRadio: ButtonRadio,
  RadioGroup: RadioGroup,
  Checkbox: Checkbox,
  ButtonCheckbox: ButtonCheckbox,
  CheckboxGroup: CheckboxGroup,
  NumberInput: NumberInput,
  FileInput: FileInput,
  DatePicker: DatePicker,
  TimePicker: TimePicker,
  TagSelector: TagSelector,
  SubmitButton: SubmitButton,
  ConfigProvider: FormConfigContext.Provider,
  Switch: Switch,
  Rating: Rating,
  Dropzone: Dropzone,
  PasswordInput: PasswordInput,
  FieldRequirements: FieldRequirements,
  RichTextEditor: RichTextEditor,
  AvatarUpload: AvatarUpload,
})
