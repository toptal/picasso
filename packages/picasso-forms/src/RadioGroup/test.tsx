import { render } from '@toptal/picasso/test-utils'
import React from 'react'

import Radio from '../Radio'
import Form from '../Form'
import type { Props } from './RadioGroup'
import RadioGroup from './RadioGroup'

const arrangeTest = ({ titleCase }: Partial<Props> = {}) =>
  render(
    <Form onSubmit={() => {}}>
      <RadioGroup
        required
        name='radio-group'
        label='Radio group label'
        titleCase={titleCase}
      >
        <Radio label='radio-label-0' value='radio-value-0' />
        <Radio label='radio-label-1' value='radio-value-1' />
      </RadioGroup>
    </Form>
  )

describe('RadioGroup', () => {
  it('shows the label in default case', () => {
    const { getByText } = arrangeTest()

    expect(getByText('Radio group label')).toBeInTheDocument()
  })

  it('shows the label in title case', () => {
    const { getByText } = arrangeTest({ titleCase: true })

    expect(getByText('Radio Group Label')).toBeInTheDocument()
  })
})
