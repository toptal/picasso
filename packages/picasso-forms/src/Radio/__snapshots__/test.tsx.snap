// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRadio renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class=""
    >
      <form
        class=""
      >
        <div
          class="PicassoFormField-root"
          data-field-has-error="false"
        >
          <label
            class="PicassoFormLabel-root"
            for="color"
          >
            <span
              class="PicassoFormLabel-medium"
            >
              What's your favorite color?
               (optional)
            </span>
          </label>
          <div
            class="MuiFormGroup-root"
            id="color"
            role="radiogroup"
            type="radio"
          >
            <div
              class="MuiGrid-root PicassoRadioGroup-grid MuiGrid-container PicassoGrid-container MuiGrid-direction-xs MuiGrid-align-items-xs-flex"
            >
              <div
                class="MuiGrid-root PicassoRadioGroup-gridItem MuiGrid-item"
              >
                <label
                  class="PicassoFormControlLabel-root Radio-root Radio-labelWithRightSpacing"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root MuiRadio-root Radio-root Radio-withLabel"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <input
                        class="PrivateSwitchBase-input"
                        name="color"
                        type="radio"
                        value="#ed143d"
                      />
                      <div
                        class="Radio-uncheckedIcon"
                      />
                    </span>
                  </span>
                  <span
                    class="PicassoFormLabel-root PicassoFormLabel-inline PicassoFormControlLabel-label Radio-label"
                  >
                    <span
                      class="PicassoFormLabel-medium"
                    >
                      Crimson
                    </span>
                  </span>
                </label>
              </div>
              <div
                class="MuiGrid-root PicassoRadioGroup-gridItem MuiGrid-item"
              >
                <label
                  class="PicassoFormControlLabel-root Radio-root Radio-labelWithRightSpacing"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root MuiRadio-root Radio-root Radio-withLabel PrivateSwitchBase-checked Mui-checked"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <input
                        checked=""
                        class="PrivateSwitchBase-input"
                        name="color"
                        type="radio"
                        value="#ffe4b5"
                      />
                      <div
                        class="Radio-checkedIcon"
                      />
                    </span>
                  </span>
                  <span
                    class="PicassoFormLabel-root PicassoFormLabel-inline PicassoFormControlLabel-label Radio-label"
                  >
                    <span
                      class="PicassoFormLabel-medium"
                    >
                      Moccasin
                    </span>
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`FormRadio required with asterisk 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class=""
    >
      <form
        class=""
      >
        <div
          class="PicassoFormField-root"
          data-field-has-error="false"
        >
          <label
            class="PicassoFormLabel-root"
            for="color"
          >
            <span
              class="PicassoFormLabel-medium"
            >
              <span
                class="PicassoFormLabel-asterisk"
              >
                *
              </span>
              What's your favorite color?
            </span>
          </label>
          <div
            class="MuiFormGroup-root"
            id="color"
            role="radiogroup"
            type="radio"
          >
            <div
              class="MuiGrid-root PicassoRadioGroup-grid MuiGrid-container PicassoGrid-container MuiGrid-direction-xs MuiGrid-align-items-xs-flex"
            >
              <div
                class="MuiGrid-root PicassoRadioGroup-gridItem MuiGrid-item"
              >
                <label
                  class="PicassoFormControlLabel-root Radio-root Radio-labelWithRightSpacing"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root MuiRadio-root Radio-root Radio-withLabel"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <input
                        class="PrivateSwitchBase-input"
                        name="color"
                        type="radio"
                        value="#ed143d"
                      />
                      <div
                        class="Radio-uncheckedIcon"
                      />
                    </span>
                  </span>
                  <span
                    class="PicassoFormLabel-root PicassoFormLabel-inline PicassoFormControlLabel-label Radio-label"
                  >
                    <span
                      class="PicassoFormLabel-medium"
                    >
                      Crimson
                    </span>
                  </span>
                </label>
              </div>
              <div
                class="MuiGrid-root PicassoRadioGroup-gridItem MuiGrid-item"
              >
                <label
                  class="PicassoFormControlLabel-root Radio-root Radio-labelWithRightSpacing"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root MuiRadio-root Radio-root Radio-withLabel PrivateSwitchBase-checked Mui-checked"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <input
                        checked=""
                        class="PrivateSwitchBase-input"
                        name="color"
                        type="radio"
                        value="#ffe4b5"
                      />
                      <div
                        class="Radio-checkedIcon"
                      />
                    </span>
                  </span>
                  <span
                    class="PicassoFormLabel-root PicassoFormLabel-inline PicassoFormControlLabel-label Radio-label"
                  >
                    <span
                      class="PicassoFormLabel-medium"
                    >
                      Moccasin
                    </span>
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
